src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,468 @@
+    std::string const user_agent = settings.get_bool(settings_pack::anonymous_mode)
+			&& !req.private_torrent ? "" : settings.get_str(settings_pack::user_agent);
+	m_websocket->set_user_agent(user_agent);
+
+#ifndef TORRENT_DISABLE_LOGGING
+	if (auto cb = requester())
+		cb->debug_log("*** WEBSOCKET_TRACKER_CONNECT [ url: %s ]", req.url.c_str());
+#endif
+
+	ADD_OUTSTANDING_ASYNC("websocket_tracker_connection::on_connect");
+	m_websocket->async_connect(req.url, std::bind(&websocket_tracker_connection::on_connect
+				, this, weak_from_this(), _1));
+}
+
+void websocket_tracker_connection::close()
+{
+	if (m_websocket)
+	{
+		m_websocket->close();
+		m_websocket.reset();
+	}
+
+	error_code const ec{error::operation_aborted};
+	while (!m_pending.empty())
+	{
+		auto [msg, callback] = std::move(m_pending.front());
+		TORRENT_UNUSED(msg);
+		m_pending.pop();
+		if (auto cb = callback.lock())
+			cb->tracker_request_error(
+					tracker_req()
+                    , ec
+                    , operation_t::sock_read
+                    , ec.message()
+                    , seconds32(120));
+	}
+
+	m_callbacks.clear();
+	m_man.remove_request(this);
+}
+
+bool websocket_tracker_connection::is_started() const
+{
+	return m_websocket && (m_websocket->is_open() || m_websocket->is_connecting());
+}
+
+bool websocket_tracker_connection::is_open() const
+{
+	return m_websocket && m_websocket->is_open();
+}
+
+void websocket_tracker_connection::queue_request(tracker_request req, std::weak_ptr<request_callback> cb)
+{
+	m_pending.emplace(tracker_message{std::move(req)}, cb);
+	if (is_open()) send_pending();
+}
+
+void websocket_tracker_connection::queue_answer(tracker_answer ans)
+{
+	m_pending.emplace(tracker_message{std::move(ans)}, std::weak_ptr<request_callback>{});
+	if (is_open()) send_pending();
+}
+
+void websocket_tracker_connection::send_pending()
+{
+	if (m_sending || m_pending.empty())
+		return;
+
+	m_sending = true;
+
+	tracker_message msg;
+    std::weak_ptr<request_callback> callback;
+    std::tie(msg, callback) = std::move(m_pending.front());
+	m_pending.pop();
+
+	std::visit([&](auto const& m)
+		{
+			// Update requester and store callback
+			if (callback.lock())
+			{
+				m_requester = callback;
+				m_callbacks[m.info_hash] = callback;
+			}
+
+			do_send(m);
+		}
+		, msg);
+}
+
+void websocket_tracker_connection::do_send(tracker_request const& req)
+{
+	if (!is_open()) return;
+
+	// Update request
+	m_req = req;
+
+	json::object payload;
+	payload["action"] = "announce";
+	payload["info_hash"] = latin1_utf8(req.info_hash);
+	payload["uploaded"] = req.uploaded;
+	payload["downloaded"] = req.downloaded;
+	payload["left"] = req.left;
+	payload["corrupt"] = req.corrupt;
+	payload["numwant"] = req.num_want;
+
+	char str_key[9];
+	std::snprintf(str_key, sizeof(str_key), "%08X", req.key);
+	payload["key"] = str_key;
+
+	static const char* event_string[] = {"completed", "started", "stopped", "paused"};
+	if (req.event != event_t::none)
```

Comment:
maybe there should be an assert that `req.event` is within the bounds of the array it's used to look up into